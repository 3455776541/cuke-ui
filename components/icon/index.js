import React from "react";
import {
	FiVolume2,
	FiLoader,
	FiCheckCircle,
	FiXCircle,
	FiAlertCircle,
	FiX,
	FiChevronUp,
	FiArrowDown
} from "react-icons/fi";

export const InfoIcon = props => <FiVolume2 {...props} />;
export const LoadingIcon = props => <FiLoader {...props} />;
export const SuccessIcon = props => <FiCheckCircle {...props} />;
export const ErrorIcon = props => <FiXCircle {...props} />;
export const WarningIcon = props => <FiAlertCircle {...props} />;
export const CloseIcon = props => <FiX {...props} />;
export const UpIcon = props => <FiChevronUp {...props} />;
export const DownIcon = props => <FiArrowDown {...props} />;
