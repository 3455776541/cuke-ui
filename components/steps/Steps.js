import React, { PureComponent, cloneElement } from "react";
import cls from "classnames";
import PropTypes from "prop-types";

export default class Steps extends PureComponent {
  state = {
    activeKey: ~~(this.props.activeKey || this.props.defaultActiveKey),
    lineWidth: 0,
    lineOffsetLeft: 0
  };
  static defaultProps = {
    prefixCls: "cuke-steps",
    defaultActiveKey: "1",
    onChange: () => {}
  };

  static propTypes = {
    prefixCls: PropTypes.string.isRequired,
    defaultActiveKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    activeKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    tabBarExtraContent: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.object
    ]),
    onChange: PropTypes.func
  };
  constructor(props) {
    super(props);
    this.tabsHeader = React.createRef();
  }

  componentWillReceiveProps({ activeKey }) {
    const _activeKey = ~~activeKey;
    if (_activeKey !== this.props.activeKey) {
      this.setState({ activeKey: _activeKey });
    }
  }

  setActiveLineStyle = () => {
    const activeElement = this.activeTab;
    const { width, left } = activeElement.getBoundingClientRect();
    const {
      left: headerOffset
    } = this.tabsHeader.current.getBoundingClientRect();
    this.setState({
      lineWidth: width,
      lineOffsetLeft: left - headerOffset
    });
  };
  onTabChange = key => {
    this.setState({ activeKey: key }, () => {
      this.setActiveLineStyle();
    });
    this.props.onChange(key);
  };

  render() {
    const {
      prefixCls,
      className,
      tabBarExtraContent,
      children,
      activeKey: _activeKey, //eslint-disable-line
      defaultActiveKey, //eslint-disable-line
      ...attr
    } = this.props;

    const { activeKey } = this.state;

    const content = React.Children.map(children, (element, index) => {
      const key = (index + 1) >> 0;
      return cloneElement(element, {
        activeKey,
        visible: activeKey === key,
        key: index
      });
    });

    const header = React.Children.map(
      children,
      ({ props: { tab, disabled } }, index) => {
        const key = (index + 1) >> 0;
        const bindActiveRef =
          activeKey === key ? { ref: node => (this[`activeTab`] = node) } : {};
        return (
          <div
            key={index}
            role="tab"
            aria-disabled={false}
            aria-selected={true}
            className={cls(`${prefixCls}-tab`, {
              [`${prefixCls}-tab-active`]: activeKey === key,
              [`${prefixCls}-tab-disabled`]: !!disabled
            })}
            {...bindActiveRef}
            onClick={() => !disabled && this.onTabChange(key)}
          >
            {tab}
          </div>
        );
      }
    );

    return (
      <div className={cls(prefixCls, className)} {...attr}>
        <div className={cls(`${prefixCls}-header`)} ref={this.tabsHeader}>
          {header}
          {tabBarExtraContent ? (
            <div className={cls(`${prefixCls}-extra`)}>
              {tabBarExtraContent}
            </div>
          ) : (
            undefined
          )}
        </div>
        <div className={cls(`${prefixCls}-content`)}>{content}</div>
      </div>
    );
  }
}
