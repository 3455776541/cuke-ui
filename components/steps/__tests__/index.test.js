import React from "react";
import { render, shallow } from "enzyme";
import assert from "power-assert";
import toJson from "enzyme-to-json";
import Steps from "../index";
import { SuccessIcon, LoadingIcon, ErrorIcon } from "../../icon";
import Spin from "../../spin";

describe("<Steps/>", () => {
  it("should render Steps", () => {
    const wrapper = render(
      <Steps>
        <Steps.Step title="第一步" description="打开冰箱111111">
          1
        </Steps.Step>
        <Steps.Step title="第二步" description="放入大象22222">
          2
        </Steps.Step>
        <Steps.Step title="第三步" description="关上冰箱33333">
          3
        </Steps.Step>
      </Steps>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should find cuke-steps class name", () => {
    const wrapper = shallow(
      <Steps>
        <Steps.Step title="第一步" description="打开冰箱111111">
          1
        </Steps.Step>
        <Steps.Step title="第二步" description="放入大象22222">
          2
        </Steps.Step>
        <Steps.Step title="第三步" description="关上冰箱33333">
          3
        </Steps.Step>
      </Steps>
    );
    assert(wrapper.find(".cuke-steps").length === 1);
    assert(wrapper.find(".cuke-steps-step").length === 3);
  });

  it("should find process classnames", () => {
    const wrapper = shallow(
      <Steps current={1}>
        <Steps.Step title="第一步" description="打开冰箱111111">
          1
        </Steps.Step>
        <Steps.Step title="第二步" description="放入大象22222">
          2
        </Steps.Step>
        <Steps.Step title="第三步" description="关上冰箱33333">
          3
        </Steps.Step>
      </Steps>
    );
    assert(wrapper.find(".cuke-steps-step-process").length === 1);
  });

  it("should find error classnames when status is error", () => {
    const wrapper = shallow(
      <Steps current={1} status="error">
        <Steps.Step title="第一步" description="打开冰箱111111">
          1
        </Steps.Step>
        <Steps.Step title="第二步" description="放入大象22222">
          2
        </Steps.Step>
        <Steps.Step title="第三步" description="关上冰箱33333">
          3
        </Steps.Step>
      </Steps>
    );
    assert(wrapper.find(".cuke-steps-step-error").length === 1);
  });

  it("should find error icon when status is error", () => {
    const wrapper = shallow(
      <Steps current={1} status="error">
        <Steps.Step title="第一步" description="打开冰箱111111">
          1
        </Steps.Step>
        <Steps.Step title="第二步" description="放入大象22222">
          2
        </Steps.Step>
        <Steps.Step title="第三步" description="关上冰箱33333">
          3
        </Steps.Step>
      </Steps>
    );
    assert(wrapper.find(ErrorIcon).length === 1);
  });
  it("should find wait icon when status is error", () => {
    const wrapper = shallow(
      <Steps current={1} status="wait">
        <Steps.Step title="第一步" description="打开冰箱111111">
          1
        </Steps.Step>
        <Steps.Step title="第二步" description="放入大象22222">
          2
        </Steps.Step>
        <Steps.Step title="第三步" description="关上冰箱33333">
          3
        </Steps.Step>
      </Steps>
    );
    assert(wrapper.find(SuccessIcon).length >= 1);
  });
  it("should render description", () => {
    const wrapper = shallow(
      <Steps current={1} status="wait">
        <Steps.Step title="第一步" description="打开冰箱111111">
          1
        </Steps.Step>
        <Steps.Step title="第二步" description="放入大象22222">
          2
        </Steps.Step>
        <Steps.Step title="第三步" description="关上冰箱33333">
          3
        </Steps.Step>
      </Steps>
    );
    assert(wrapper.find(".cuke-steps-step-content-description").length >= 1);
  });

  it("should can not render description", () => {
    const wrapper = shallow(
      <Steps current={1} status="wait">
        <Steps.Step title="第一步">1</Steps.Step>
        <Steps.Step title="第二步">2</Steps.Step>
        <Steps.Step title="第三步">3</Steps.Step>
      </Steps>
    );
    assert(wrapper.find(".cuke-steps-step-content-description").length === 0);
  });

  it("should render custom icon", () => {
    const wrapper = shallow(
      <Steps current={1}>
        <Steps.Step title="第一步" icon={<SuccessIcon />}>
          1
        </Steps.Step>
        <Steps.Step title="第二步" icon={<ErrorIcon />}>
          2
        </Steps.Step>
        <Steps.Step title="第三步" icon={<LoadingIcon />}>
          3
        </Steps.Step>
      </Steps>
    );
    assert(wrapper.find(SuccessIcon).length === 1);
    assert(wrapper.find(ErrorIcon).length === 1);
    assert(wrapper.find(LoadingIcon).length === 1);
  });

  it("should render spin when showProcessSpin is true", () => {
    const wrapper = shallow(
      <Steps current={1} showProcessSpin>
        <Steps.Step title="第一步">1</Steps.Step>
        <Steps.Step title="第二步">2</Steps.Step>
        <Steps.Step title="第三步">3</Steps.Step>
      </Steps>
    );
    assert(wrapper.find(Spin).length === 1);
  });

  it("should can not render spin when showProcessSpin is false", () => {
    const wrapper = shallow(
      <Steps current={1} showProcessSpin={false}>
        <Steps.Step title="第一步">1</Steps.Step>
        <Steps.Step title="第二步">2</Steps.Step>
        <Steps.Step title="第三步">3</Steps.Step>
      </Steps>
    );
    assert(wrapper.find(Spin).length === 0);
  });

  it("should find two is done status classnames", () => {
    const wrapper = shallow(
      <Steps current={1}>
        <Steps.Step title="第一步">1</Steps.Step>
        <Steps.Step title="第二步">2</Steps.Step>
        <Steps.Step title="第三步">3</Steps.Step>
      </Steps>
    );
    wrapper.setProps(
      {
        current: 2
      },
      () => {
        assert(wrapper.find(".cuke-steps-step-done").length === 2);
      }
    );
  });
});
