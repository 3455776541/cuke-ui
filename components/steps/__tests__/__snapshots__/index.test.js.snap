// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Steps/> should render Steps 1`] = `
<div
  class="cuke-steps"
>
  <div
    class="cuke-steps-header"
  >
    <div
      aria-selected="true"
      class="cuke-steps-step cuke-steps-step-active cuke-steps-step-process"
      role="step"
    >
      <div
        class="cuke-steps-step-icon"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <line
            x1="12"
            x2="12"
            y1="2"
            y2="6"
          />
          <line
            x1="12"
            x2="12"
            y1="18"
            y2="22"
          />
          <line
            x1="4.93"
            x2="7.76"
            y1="4.93"
            y2="7.76"
          />
          <line
            x1="16.24"
            x2="19.07"
            y1="16.24"
            y2="19.07"
          />
          <line
            x1="2"
            x2="6"
            y1="12"
            y2="12"
          />
          <line
            x1="18"
            x2="22"
            y1="12"
            y2="12"
          />
          <line
            x1="4.93"
            x2="7.76"
            y1="19.07"
            y2="16.24"
          />
          <line
            x1="16.24"
            x2="19.07"
            y1="7.76"
            y2="4.93"
          />
        </svg>
      </div>
      <div
        class="cuke-steps-step-content"
      >
        <div
          class="cuke-steps-step-content-title"
        >
          第一步
        </div>
        <p
          class="cuke-steps-step-content-description"
        >
          打开冰箱111111
        </p>
      </div>
    </div>
    <div
      aria-selected="true"
      class="cuke-steps-step cuke-steps-step-wait"
      role="step"
    >
      <div
        class="cuke-steps-step-icon"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
          />
          <polyline
            points="22 4 12 14.01 9 11.01"
          />
        </svg>
      </div>
      <div
        class="cuke-steps-step-content"
      >
        <div
          class="cuke-steps-step-content-title"
        >
          第二步
        </div>
        <p
          class="cuke-steps-step-content-description"
        >
          放入大象22222
        </p>
      </div>
    </div>
    <div
      aria-selected="true"
      class="cuke-steps-step cuke-steps-step-wait"
      role="step"
    >
      <div
        class="cuke-steps-step-icon"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
          />
          <polyline
            points="22 4 12 14.01 9 11.01"
          />
        </svg>
      </div>
      <div
        class="cuke-steps-step-content"
      >
        <div
          class="cuke-steps-step-content-title"
        >
          第三步
        </div>
        <p
          class="cuke-steps-step-content-description"
        >
          关上冰箱33333
        </p>
      </div>
    </div>
  </div>
  <div
    class="cuke-steps-content"
  />
</div>
`;
